Subscribers Dataview: add sort order

arcangelini opened the pull request:
Closes https://github.com/Automattic/loop/issues/303
Closes https://github.com/Automattic/loop/issues/305
Closes https://github.com/Automattic/loop/issues/349

## Proposed Changes

* Adds sort order to the subscriber dataviews
* Extends the subscription context to include the sort order
* Update the query args on sort order update

## Why are these changes being made?

* This is needed to accommodate the sorting in data views

## Testing Instructions

1. Open live link and go to `/subscribers?flags=subscribers-dataviews`
2. Sort by name or date ascending and descending
3. Remove flag and test for regressions

allilevine reviewed (commented):
Looks good overall! Two things:

1) I can click Subscription Type to sort, but I think it's actually using the Since column. Did you want to add that sorting in this PR?
2) When I first sort, I see the loading circle for a millisecond before the subscribers load. I'm not sure why this would be any slower than what's on staging now, though. An extra check for the sort order?

arcangelini commented:
> I can click Subscription Type to sort, but I think it's actually using the Since column. Did you want to add that sorting in this PR?

I removed this sort until we extend the API. I also added more logic to prevent this from defaulting to date sort like you were seeing. Just in case.

> When I first sort, I see the loading circle for a millisecond before the subscribers load. I'm not sure why this would be any slower than what's on staging now, though. An extra check for the sort order?

Well, I am not sure I see that but as of right now I am seeing the dot persist across the headers. For example, click DATE ⇢ ASC and you will see the dot move. Go to NAME ⇢ DESC and the dot should update, now go check DATE again and you will see the dot still showing in the date section as well as the name section.

arcangelini committed:
Rebased and added sorting and moving columns

arcangelini committed:
Fix pagination

arcangelini committed:
Merge branch 'trunk' into subscribers/dataview-sorting

arcangelini commented on client/my-sites/subscribers/components/subscriber-data-views/subscriber-data-views.tsx:
```diff
@@ -69,9 +63,22 @@ const SubscriberDataViews = ( {
+		setSortOrder,
 	} = useSubscribersPage();
 
+	const [ currentView, setCurrentView ] = useState< View >( {
+		type: 'table',
+		layout: {},
+		page,
+		perPage,
+		sort: {
+			field: sortTerm,
+			direction: sortOrder,
+		},
+	} );
```
Moving the view to state made it easier to maintain.

arcangelini commented on client/my-sites/subscribers/components/subscriber-data-views/subscriber-data-views.tsx:
```diff
@@ -149,7 +156,7 @@ const SubscriberDataViews = ( {
 				render: ( { item }: { item: Subscriber } ) => <SubscriptionTypeCell subscriber={ item } />,
 				enableHiding: false,
-				enableSorting: true,
+				enableSorting: false,
```
I will add this back after we adjust the API to filter by these.

arcangelini commented on client/my-sites/subscribers/components/subscriber-data-views/subscriber-data-views.tsx:
```diff
@@ -215,62 +203,57 @@ const SubscriberDataViews = ( {
+				setSortOrder( newView.sort.direction );
+			}
+
+			// Handle field order change
+			if ( newView.fields && newView.fields !== currentView.fields ) {
+				setCurrentView( ( oldCurrentView ) => ( {
+					...oldCurrentView,
+					fields: newView.fields,
+				} ) );
```
This allows the columns to be shifted ⇠ ⇢

arcangelini commented on client/my-sites/subscribers/components/subscriber-data-views/subscriber-data-views.tsx:
```diff
@@ -281,6 +264,35 @@ const SubscriberDataViews = ( {
 		};
 	}, [ subscribers, grandTotal, pages ] );
 
+	// Update the view when a subscriber is selected
+	useEffect( () => {
+		const commonViewProps = {
+			page,
+			perPage,
+			sort: {
+				field: sortTerm,
+				direction: sortOrder,
+			},
+		};
+
+		setCurrentView( ( oldCurrentView ) => ( {
+			...oldCurrentView,
+			...commonViewProps,
+			type: selectedSubscriber ? 'list' : 'table',
+			layout: selectedSubscriber
+				? {
+						showMedia: true,
+						mediaSize: 40,
+						mediaField: 'media',
+						primaryField: 'name',
+				  }
+				: {},
+			fields: selectedSubscriber
+				? [ 'media', 'name' ]
+				: [ 'name', ...( ! isMobile ? [ 'subscription_type', 'date_subscribed' ] : [] ) ],
+		} ) );
+	}, [ isMobile, selectedSubscriber, page, perPage, sortTerm, sortOrder ] );
```
Centralizing the view state made it easier to read and understand what was changing.

arcangelini commented on client/my-sites/subscribers/components/subscribers-page/subscribers-page-context.tsx:
```diff
@@ -77,7 +80,12 @@ export const SubscribersPageProvider = ( {
 }: SubscribersPageProviderProps ) => {
+	const isDataView = isEnabled( 'subscribers-dataviews' );
 	const [ perPage, setPerPage ] = useState( DEFAULT_PER_PAGE );
+	const [ sortOrder, setSortOrder ] = useState< 'asc' | 'desc' >( 'desc' );
+	const [ dataViewSortTerm, setDataViewSortTerm ] = useState< SubscribersSortBy >(
+		SubscribersSortBy.DateSubscribed
+	);
```
Rather than adding these to the query string watcher I just placed this in state

arcangelini commented on client/my-sites/subscribers/components/subscribers-page/subscribers-page-context.tsx:
```diff
@@ -219,8 +228,10 @@ export const SubscribersPageProvider = ( {
 				pageChangeCallback,
-				sortTerm,
-				setSortTerm: sortTermChanged,
+				sortTerm: isDataView ? dataViewSortTerm : sortTerm,
+				setSortTerm: isDataView ? setDataViewSortTerm : sortTermChanged,
```
In order to not affect the current version of this I only use the state for dataViews

arcangelini commented on client/my-sites/subscribers/queries/use-subscribers-query.tsx:
```diff
@@ -48,10 +51,17 @@ const useSubscribersQuery = ( {
 					? SubscribersFilterBy.WPCOM
 					: filterOption;
 
+			const params = new URLSearchParams( {
+				per_page: perPage.toString(),
+				page: page.toString(),
+				[ userTypeField ]: validatedFilterOption,
+				...( search && { search } ),
+				...( sortTerm && { sort: sortTerm } ),
+				...( sortOrder && { sort_order: sortOrder } ),
+			} );
+
```
This was easier to read IMO

arcangelini committed:
Fix type test

allilevine reviewed (commented):
Thanks for centralizing the view logic! 🙌 The only thing I spotted was that the Name sort order was flipped in the non-dataviews view. When I sort by name it starts with the end of the alphabet.

arcangelini committed:
Merge branch 'trunk' into subscribers/dataview-sorting

arcangelini committed:
Fix sort direction in prod

arcangelini committed:
Merge branch 'trunk' into subscribers/dataview-sorting

allilevine reviewed (approved):
The sort order without the flag is correct. ✅
